core/src/main/java/org/apache/iceberg/util/ParallelIterable.java

```diff
@@ -136,19 +152,20 @@ private boolean checkTasks() {
         }
       }
 
-      return !closed && (tasks.hasNext() || hasRunningTask);
+      return !closed.get() && (tasks.hasNext() || hasRunningTask);
     }
 
-    private Future<?> submitNextTask() {
-      if (!closed && tasks.hasNext()) {
-        return workerPool.submit(tasks.next());
+    private Future<Optional<Task<T>>> submitNextTask() {
+      if (!closed.get() && (!yieldedTasks.isEmpty() || tasks.hasNext())) {
+        return workerPool.submit(
+            !yieldedTasks.isEmpty() ? yieldedTasks.removeFirst() : tasks.next());
```

Comment:
If both `yieldedTasks.removeFirst()` and `tasks.next()` are actually called and executed then some results will be discarded. That means correctness here depends on whether the ternary operator will actually execute the expression that is not used.

While this _is_ correct and the unused expression is not executed, I generally prefer not to rely on behavior like that since it is hard to remember. I think it is better to use a standard `if`/`else` statement:

```java
  if (yieldedTasks.isEmpty()) {
    return workerPool.submit(tasks.next());
  } else {
    return workerPool.submit(yieldedTasks.removeFirst());
  }
```

Reply:
done!